### DISCUSSION TITLE
React CreateAndApplyTexture

### DISCUSSION THREAD

**Discussion by adrianposteuca** [google/model-viewer#3975] [General]
Hi all,

I am building a react app with ModelViwer, But ai want to access model materials and use CreateAndApplytexture , but I Can not access Modelviwer functions like CreateTexture or Model.materials

Any help is much appreciated!

I added the Modelviwer core

` </div>
      <model-viewer className ="modelViewer"
      id ="modelViewer"
      alt="Shirt"
      scr = {`${modelGLB}`}
      //src="https://cdn.glitch.global/6430467c-091d-4527-99e0-5d9c034eac15/tshirt_test.glb?v=1669301875394app\src\Assets\tshirt_test.glb"
      shadow-intensity="1"
      camera-controls
      touch-action="pan-y"
      ref= {modelRef}
    ></model-viewer>`

But I can not access materials " const material = modelRef.current.materials[0];"

How can I create  this exemple

`const createAndApplyTexture = async (channel, event) => {
      const texture = await modelViewer.createTexture(event);
      if (channel.includes("base") || channel.includes("metallic")) {
        material.pbrMetallicRoughness[channel].setTexture(texture);
      } else {
        material[channel].setTexture(texture);
      }
    };
createAndApplyTexture("baseColorTexture", myimage)
`

**Comment by elalish** [comment]
Did you mean `const material = modelRef.current.model.materials[0];`?

**Reply by adrianposteuca** [reply]
Yes... it returns an error. it can not access materials

Uncaught TypeError: Cannot read properties of undefined (reading 'model')

**Reply by elalish** [reply]
Sounds like your `modelRef` isn't working. I'm not a React user, so I generally get the element via `querySelector`. Can you debug enough that you at least get an element back?

**Reply by adrianposteuca** [reply]
But can I use the function CreateTexture? Inside react?

const createAndApplyTexture = async (channel, event) => {
      const texture = await modelViewer.createTexture(event);
      if (channel.includes("base") || channel.includes("metallic")) {
        material.pbrMetallicRoughness[channel].setTexture(texture);
      } else {
        material[channel].setTexture(texture);
      }
    };

**Reply by elalish** [reply]
It should work if you've found the model-viewer element.

**Reply by adrianposteuca** [reply]
@elalish 
This is my object as ref ={modelviwer}..Console.log{ "myContainer..." +ref )
But I can not access ref.current.model.materials[0]; => cannot read the properties of undefined

 "myContainer.. 
<model-viewer src="./tshirt.glb" alt="A 3D model of an astronaut" shadow-intensity="1" camera-controls="true" auto-rotate="true" ar="true" ar-status="not-presenting"></model-viewer>"

**Reply by akshays-repo** [reply]
Hi, @adrianposteuca it's 2024 and I'm facing the same issue can you help me with this ??? how do you crack this issue?

**Reply by elalish** [reply]
Also, make sure you've waited for everything to finish loading - the MV element will only have a `model` API once the GLB has finished loading.

**Reply by akshays-repo** [reply]
```
useEffect(() => {
     
            const checkModelReady = () => {
                return new Promise(resolve => {
                    const check = () => {
                        if (modelRef.current.loaded) {
                            resolve();
                        } else {
                            setTimeout(check, 100);
                        }
                    };
                    check();
                });
            };

            await checkModelReady();
            
            // rest of the code goes here

        
    }, []);
```

I did something similar and it's working for me

**Reply by elalish** [reply]
A simpler approach:
```
modelRef.current.addEventListener('load', () => {// rest of the code goes here});
```

**Reply by akshays-repo** [reply]
Hi, thanks for the support. I'm exploring this package for my side project. So, my idea is to display a list of paintings. When a user clicks on one, we can view it on the wall. This is the concept. For this, I'm using a single GLB model frame file and adding an image as the texture. I saw the texture change example on the site, and those are working fine. However, in my case, it's not working at all. I'm getting a frame model without the photo. Also, while exploring the texture example on the website, it opens in the browser itself. However, my code is opening it in the Google app since I'm using Android

```
import React, { useEffect, useRef, useState } from 'react'
import '@google/model-viewer/';

const TextureModel = () => {
    const modelRef = useRef()

    useEffect(() => {
        const applyTextures = async () => {
            const imageTextureUrl = "/phtotframe/Textures/Image.JPG";
            modelRef?.current?.addEventListener('load', async() => {
            const imageTexture = await modelRef.current.createTexture(imageTextureUrl);
            imageTexture.name = "image_texture";
            const model = modelRef.current.model;
            const material2 = modelRef.current.model.materials[3];
            material2.pbrMetallicRoughness.baseColorTexture.setTexture(imageTexture);
            })
        };

        applyTextures();
    }, []);


    return (
        <div>
            <model-viewer 
                id="frame"
                camera-controls 
                touch-action="pan-y" 
                ref={modelRef}
                src={'/phtotframe/model/frame1.glb'}
                ar-placement="wall"
                tone-mapping="commerce" 
                ar 
                alt="A 3D model of a Shoe">
            </model-viewer>
        </div>
    )
}

export default TextureModel;

```

**Reply by akshays-repo** [reply]
![MicrosoftTeams-image (4)](https://github.com/google/model-viewer/assets/49100604/f3b54335-25fc-40cc-ab1c-6b9456690e32)
![MicrosoftTeams-image (3)](https://github.com/google/model-viewer/assets/49100604/5e6c312b-2c62-4fb8-90c6-c7359a7e9246)
check this output is I missing anything ?? or doing something anti-practice

**Reply by elalish** [reply]
Try `ar-modes="webxr quick-look"` - scene-viewer won't work for modified GLBs. If WebXR isn't working, you'll need to check your console to see why. Make sure you're serving over HTTPS. If you're still having trouble, see if it's this https://github.com/google/model-viewer/issues/4582 which appears to be device-specific.
